graylog2-web-interface/src/components/grok-patterns/GrokPatternInput.jsx

```diff
@@ -0,0 +1,137 @@
+
+class GrokPatternInput extends React.Component {
+  static propTypes = {
+    pattern: PropTypes.string,
+    patterns: PropTypes.array,
+    onPatternChange: PropTypes.func,
+  };
+
+  static defaultProps = {
+    pattern: '',
+    patterns: [],
+    onPatternChange: () => {},
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      patternFilter: '',
+      activeListItem: -1,
+    };
+  }
+
+  shownListItems = [];
+
+  _onPatternChange = (e) => {
+    this.props.onPatternChange(e.target.value);
+  };
+
+  _onPatternFilterChange = (e) => {
+    this.setState({ patternFilter: e.target.value, activeListItem: -1 });
+  };
+
+  _onPatternFilterKeyDown = (e) => {
+    const ARROW_DOWN = 40;
+    const ARROW_UP = 38;
+    const ENTER = 13;
+
+    let activeListItem = 0;
+    let domElement;
+    switch (e.keyCode) {
+      case ARROW_DOWN:
+        activeListItem = this.state.activeListItem + 1;
+        if (activeListItem >= this.shownListItems.length) {
+          return;
+        }
+        domElement = document.getElementById(`list-item-${activeListItem}`);
+        domElement.scrollIntoView({ behavior: 'smooth' });
+        this.setState({ activeListItem: activeListItem });
+        e.preventDefault();
+        break;
+      case ARROW_UP:
+        activeListItem = this.state.activeListItem - 1;
+        if (activeListItem < 0) {
+          return;
+        }
+        domElement = document.getElementById(`list-item-${activeListItem}`);
+        domElement.scrollIntoView();
+        this.setState({ activeListItem: activeListItem });
+        e.preventDefault();
+        break;
+      case ENTER:
+        this._addToPattern(this.shownListItems[this.state.activeListItem]);
+        e.preventDefault();
+        break;
+      default:
+        break;
+    }
+  };
+
+  _addToPattern = (name) => {
+    const pattern = this.props.pattern || '';
+    const index = this.patternInput.getInputDOMNode().selectionStart || pattern.length;
+    const newPattern = `${pattern.slice(0, index)}%{${name}}${pattern.slice(index)}`;
+    this.props.onPatternChange(newPattern);
+  };
+
+  render() {
+    const regExp = RegExp(this.state.patternFilter, 'i');
+    const maxLength = 20;
+    this.shownListItems = [];
+    const patternsToDisplay = this.props.patterns.filter(pattern => regExp.test(pattern.name))
+      .map((pattern, index) => {
+        const patternDisplay = pattern.pattern.length > maxLength ?
+          `${pattern.pattern.substring(0, maxLength)} ...` : pattern.pattern;
+        const active = index === this.state.activeListItem;
+        this.shownListItems.push(pattern.name);
+        return (
+          <ListGroupItem id={`list-item-${index}`}
+                         header={pattern.name}
+                         bsStyle={active ? 'info' : undefined}
+                         onKeyDown={this._onPatternFilterKeyDown}
+                         key={pattern.name}>
+            <span>{patternDisplay}</span>
```

Comment:
I would use the full pattern in here, and then let CSS take care of hiding the content that is not visible with:

```
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    display: block;
```

With that all patterns will use all available space when rendering on the screen, and displaying ellipsis when the content is too large for that.